src/helpers/MiscFunctions.cpp

```diff
@@ -305,7 +306,91 @@ int getWorkspaceIDFromString(const std::string& in, std::string& outName) {
+                    // Can't jump to this workspace
+                    invalidWSes.insert(ws->m_iID);
+                }
+            }
+            for (auto& rule : g_pConfigManager->getAllWorkspaceRules()) {
+                const auto PMONITOR = g_pCompositor->getMonitorFromName(rule.monitor);
+                if (!PMONITOR || PMONITOR->ID == g_pCompositor->m_pLastMonitor->ID) {
+                    // Can't be invalid
+                    continue;
+                }
+                // WS is bound to another monitor, can't jump to this
+                invalidWSes.insert(rule.workspaceId);
+            }
+
+            // Just take a blind guess at where we'll probably end up
+            int  predictedWSID = g_pCompositor->m_pLastMonitor->activeWorkspace + remains;
+            int  remainingWSes = 0;
+            char walkDir       = in[1];
+            if (predictedWSID <= 0) {
+                // We overshot, get first valid
+                predictedWSID = 0;
+                walkDir       = '+';
+                remainingWSes = 1;
+            } else {
+                // Count how many invalidWSes are in between (how bad the prediction was)
+                int  beginID = in[1] == '+' ? g_pCompositor->m_pLastMonitor->activeWorkspace + 1 : predictedWSID;
+                int  endID   = in[1] == '+' ? predictedWSID : g_pCompositor->m_pLastMonitor->activeWorkspace;
+                auto begin   = invalidWSes.upper_bound(beginID - 1); // upper_bound is >, we want >=
+                for (auto it = begin; *it <= endID && it != invalidWSes.end(); it++) {
+                    remainingWSes++;
+                }
+            }
+
+            // Go in the search direction for remainingWSes
+            // This takes at most 2 * invalidWSes.size() steps(A workspace cannot be both behind predictedWSID and after predictedWSID.
+            // This is because the loop takes at most ([number of invalidWSes between beginID and endID] + [number of invalidWSes after/before endID]) steps and it is potentially run twice)
+            int finalWSID = predictedWSID;
+            if (walkDir == '-') {
+                int beginID = finalWSID;
+                int curID   = finalWSID;
+                while (--curID > 0 && remainingWSes > 0) {
+                    if (invalidWSes.find(curID) == invalidWSes.end()) {
+                        remainingWSes--;
+                    }
+                    finalWSID = curID;
+                }
+                if (invalidWSes.find(finalWSID) != invalidWSes.end()) {
+                    // Couldn't find valid workspace in negative direction, search last first one back up positive direction
+                    walkDir = '+';
+                    // We know, that everything less than beginID is invalid, so don't bother with that
+                    finalWSID     = beginID;
+                    remainingWSes = 1;
+                }
+            }
+            if (walkDir == '+') {
```

Comment:
else if

Reply:
Nope, that's intentional, since we modify walkDir to '+' in l. 372